pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -248,32 +347,51 @@ func (r *Reconciler) ValidateHardware(ctx context.Context, log logr.Logger, tink
 	clusterSpec := tinkerbellScope.ClusterSpec
 	log = log.WithValues("phase", "validateHardware")
 
-	capiCluster, err := controller.GetCAPICluster(ctx, r.client, clusterSpec.Cluster)
-	if err != nil {
-		return controller.Result{}, errors.Wrap(err, "validating tinkerbell hardware")
-	}
-	if capiCluster != nil {
-		// If CAPI cluster exists, the hardware has been validated
-		// and it's possibly already in use so no need to validate it again.
-		log.V(3).Info("CAPI cluster already exists, skipping hardware validations")
-		return controller.Result{}, nil
-	}
-
 	// We need a new reader each time so that the catalogue gets recreated.
 	kubeReader := hardware.NewKubeReader(r.client)
 	if err := kubeReader.LoadHardware(ctx); err != nil {
 		log.Error(err, "Hardware validation failure")
 		failureMessage := err.Error()
 		clusterSpec.Cluster.Status.FailureMessage = &failureMessage
-
 		return controller.ResultWithReturn(), nil
 	}
 
 	var v tinkerbell.ClusterSpecValidator
-	v.Register(
-		tinkerbell.MinimumHardwareAvailableAssertionForCreate(kubeReader.GetCatalogue()),
-		tinkerbell.HardwareSatisfiesOnlyOneSelectorAssertion(kubeReader.GetCatalogue()),
-	)
+	v.Register(tinkerbell.HardwareSatisfiesOnlyOneSelectorAssertion(kubeReader.GetCatalogue()))
+
+	o, err := r.DetectOperation(ctx, log, tinkerbellScope)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	switch o {
+	case K8sVersionUpgradeOperation:
+		v.Register(tinkerbell.ExtraHardwareAvailableAssertionForRollingUpgrade(kubeReader.GetCatalogue()))
+	case NewClusterOperation:
+		v.Register(tinkerbell.MinimumHardwareAvailableAssertionForCreate(kubeReader.GetCatalogue()))
+	case ScaleOperation:
+		currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+		if err != nil {
+			return controller.Result{}, err
+		}
+		var wgs []*clusterapi.WorkerGroup[*tinkerbellv1.TinkerbellMachineTemplate]
+		for _, wnc := range tinkerbellScope.ClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+			md := &clusterv1.MachineDeployment{}
+			mdName := clusterapi.MachineDeploymentName(tinkerbellScope.ClusterSpec.Cluster, wnc)
+			key := types.NamespacedName{Namespace: constants.EksaSystemNamespace, Name: mdName}
+			err := r.client.Get(ctx, key, md)
+			if err == nil {
```

Comment:
Same question about if there's a need to bubble up errors that are not `IsNotFound()`